Adding Direct Method APIs to OPC Publisher.

karok2m opened the pull request:
This PR adds support of IoT Edge direct methods in OPC Publisher and adds several accompanying features.

The following IoT Edge direct methods have been added along with tests for them:

* `PublishNodes_V1`: to add a set of nodes to be published for a specific `DataSetWriter`.
* `UnpublishNodes_V1`: to remove nodes from a previously configured `DataSetWriter`.
* `UnpublishAllNodes_V1`: to remove all the nodes from a previously configured `DataSetWriter`.
* `GetConfiguredEndpoints_V1`: to get list of all configured `DataSetWriters`.
* `GetConfiguredNodesOnEndpoint_V1`: to get list of configured nodes for previously configured `DataSetWriter`.
* `GetDiagnosticInfo_V1`: to get diagnostics information for all actively processed `DataSetWriters`.
* `AddOrUpdateEndpoints_V1`:  to perform a complete `published_nodes.json` update or update multiple `DataSetWriters` at once.

Other major changes introduced by this PR

* Added configurable default value for maximum nodes per dataset/subscription. (#1498)
* Added support for incremental publisher reconfiguration. (#1545)

List of all changes:

* Added infrastructure support for device methods in publisher. (#1494)
* Added configurable default value for maximum nodes per dataset/subscription. (#1498)
* Added stress test with 100 endpoints. (#1507)
* Added unit tests for direct method api & bugfixes. (#1502)
* Added direct method E2E `PublishNodes_V1` and `UnpublishNodes_V1` test. (#1508)
* Implemented response payload for `PublishNodes_V1` direct method. (#1500)
* Added response for `UnpublishNodes_V1` direct method. (#1515)
* Added persistency to published nodes file. (#1509)
* Added note about calling `PersistPublishedNodes()` method from a locked block. (#1524)
* Added `GetConfiguredNodesOnEndpoint_V1` direct method. (#1522)
* Fixed data set detection bug in GetConfiguredNodesOnEndpointAsync(). (#1532)
* Added direct method `GetConfiguredEndpoints_V1`. (#1520)
* Added `GetConfiguredNodesOnEndpoint_V1` direct method to E2E test. (#1533)
* Added documentation of OPC Publisher direct methods. (#1538)
* Added E2E tests for `GetConfiguredNodesOnEndpoint_V1` direct method. (#1536)
* Updated publisher.md with `GetConfiguredNodesOnEndpoint_V1` direct method doc. (#1541)
* Updated `GetConfiguredEndpoints_V1` in docs. (#1542)
* Added implementation of `AddOrUpdateEndpoints_V1` direct method call. (#1539)
* Implemented `UnpublishAllNodes_V1` direct method. (#1543)
* Implemented `GetDiagnosticInfo_V1` direct method. (#1544)
* Added support for incremental publisher reconfiguration. (#1545)
* `PublishNodes_V1`/`UnpublishNodes_V1`/`AddOrUpdateEndpoints_V1`/`UnpublishAllNodes_V1` nodes empty response. (#1554)

cristipogacean committed:
integrate device methods infrastructure in publisher module

karok2m committed:
Merge branch 'main' into feature/dmapi

dacolgit committed:
start adding E2E test infrastructure for dm (#1486)

* add deploment o publisher 2.5.5

* fix deployment for publisher 2.5.5

* fix image path

* add skeleton for dm test

* removed launchsettings.json

* restored empty lauchsetiings.json file

* add deploment o publisher 2.5.5

* fix deployment for publisher 2.5.5

* fix image path

* add skeleton for dm test

* removed launchsettings.json

* restored empty lauchsetiings.json file

* removed MessagingMode for deployment of publisher 2.5.5

* fix pr comments

* fix

Co-authored-by: Karapet Kostandyan <kakostan@microsoft.com>

dacolgit committed:
create e2e test  for direct method  (#1496)

* create test  for direct methdo with current publisher versione and publisher 2.5.5

* move initialization of _iotHubConnectionString and _iotHubPublisherModuleName to the constructor

* fix pr comments

cristipogacean committed:
add infrastructure support for device methods in publisher (#1494)

* add infrastructure support for DM in publisher
* implement publish/unpublish nodes DM publisher
* fixes for publishednodes to jobs conversion
* fix configuration change propagation to workers and worker supervisor
* added support for setting queue size for publish nodes
* fix heartbeat interval setting in configuration payload
* add deterministic identifiers for endpoints, DataSetGroups, DataSetWriterIds respectively sessions and subscriptions 
* unit tests for publish/unpublish nodes
* cleanup unused legacy cli default options for publisher
* defined consistent defaults for legacy cli options for publisher
* fix infinite loop when empty pn.json with UTF88 BOM header
* small bugfixes & cosmetics
* add review feedback

cristipogacean committed:
configurable default value for maximum nodes per dataset/subscription (#1498)

* add configurable default value for maximum nodes per dataset/subscription

cristipogacean committed:
Merge branch 'main' into feature/dmapi

cristipogacean committed:
fix main merge conflict

karok2m committed:
Added stress test with 100 endpoints. (#1507)

* Added stress test with 100 endpoints.
* Implemented review comments.

cristipogacean committed:
added unit tests for DM api & bugfixes (#1502)

* fix api model
* separate locking mechanism in job orchestrator
* add stubs for not implemented DMs
* fix PublishNodesRequestApiModel serialization
* fix potential sync issue

dacolgit committed:
Direct method e2e publish and unpublish test (#1508)

* start adding methods

* add direct method call to e2e test

* remoce dependency

* Add call to publish and unpublish direct methods for current and legacy (2.5.5) version in e2e test.
Fix creation of DataSetWriterId when is not defined in pn.json file

* fix when DataSetWriterId is null
Add unit test

* fix pr comments

* add project dependencies.

monikavar committed:
implementing response payload for Publish nodes direct method (#1500)

* changes for publish nodes response

* Update PublisherExtensions.cs

monikavar committed:
add response for unpublish nodes dm (#1515)

add response for Unpublishnodes direct method

karok2m committed:
Merge branch 'main' into feature/dmapi

karok2m committed:
Added persistency to published nodes file. (#1509)

* Integrated serialization changes into PublishNodesAsync and UnpublishNodesAsync.
* Fixed tests.
* Added stress test with 100 endpoints.
* Minor refactoring of stress test.
* Added fallback mechanism in PublishedNodesProvider.
* Optimized RefreshJobs() method.
* Fixed one bug and tests.
* Addressing review comments.
* Using NewtonSoftJsonSerializer to serialize updates to published nodes file.
* Fixed place of using directives.
* Added exception handling to DeserializePublishedNodes() method.
* Addressing review comments.
* Reverted back to initial job refresh mechanism that was present in UnpublishNodesAsync().
* Reverting changes to OpcNodeModel.HeartbeatInterval.
* Implemented IPublishedNodesProvider interface.
* Fixed bug of updating published nodes before reporting an error.

cristipogacean committed:
Merge branch 'main' into feature/dmapi

karok2m committed:
Added note about calling PersistPublishedNodes() method from a locked block. (#1524)

Added note about calling PersistPublishedNodes() method from a locked block.

cristipogacean committed:
small buffix (#1526)

publisher fails to start in standalone mode with autofac exception in PublishedNodesProvider
pn.json file changes are not recognized and processed anymore

dacolgit committed:
Add GetConfiguredNodesOnEndpoint dm (#1522)

* add model

* Add GetConfiguredNodesOnEndpoints DM
add unit test

* fix merge

* fix pr comments

* fix pr comments

* remove dependency

* remove duplicated models

* fix node on endpoint filtering
added more test

* use HasSameGroup to filter nodes.
fix tests

* fix merge issue

karok2m committed:
Merge branch 'main' into feature/dmapi

karok2m committed:
Fixed data set detection bug in GetConfiguredNodesOnEndpointAsync(). (#1532)

* Changed data set detection logic in GetConfiguredNodesOnEndpointAsync() method to use IsSameDataSet() instead of HasSameGroup() which also compares DataSetWriterId and DataSetPublishingInterval properties of data sets.
* Added test which queries for nodes of specific endpoints where endpoints differ only in DataSetWriterId property.

monikavar committed:
Add direct method GetConfiguredEndpoints (#1520)

* implement GetConfiguredEndpoints DM and add tests

monikavar committed:
rename model name (#1534)

* rename to publishnodedendpointapimodel

* adding datasetwritergroup and datasetpublishinginterval

dacolgit committed:
Add GetConfiguredNodesOnEndpoint dm to e2e test (#1533)

* Add GetConfiguredNodesOnEndpoint dm to e2e test

* fix pr comments

* add direct method test for multiple endpoint
add constants for dm names

* fix pr comments

karok2m committed:
Merge branch 'main' into feature/dmapi

cristipogacean committed:
add publisher direct methods doc (#1538)

* update operations manual to 2.8.1
* add publisher DM basic documentation
* fix a couple of issues

monikavar committed:
Add e2e tests for GetConfiguredNodesOnEndpoint direct method (#1536)


* adding e2e tests for getconfiguredendpoints direct method

* checking all the properties

dacolgit committed:
update publisher.md with GetConfiguredNodesOnEndpoint dm doc (#1541)

* update publisher.md with GetConfiguredNodesOnEndpoint dm doc

* add response sample

* fix pr comments

monikavar committed:
getconfiguredendpoints update in docs

monikavar committed:
Revert "getconfiguredendpoints update in docs"

This reverts commit f473dc868e5218af118b4bd40119ca34295773c4.

monikavar committed:
update GetConfiguredEndpoints_V1 in docs (#1542)

* getconfiguredendpoints update in docs

* Update publisher.md

karok2m committed:
Added implementation of AddOrUpdateEndpoints_V1 direct method call. (#1539)

* Added implementation of AddOrUpdateEndpoints_V1 direct method call.
* Null reference bug fixed in PublishedNodesProvider class.
* Fixed data set detection bug in GetConfiguredNodesOnEndpointAsync() method.
* Added tests for AddOrUpdateEndpointsAsync() method.
* Adjusting max worker count after execution of AddOrUpdateEndpointsAsync() method.
* Added test to check functionality of adding endpoints with AddOrUpdateEndpointsAsync() method.
* Added endpoint removal tests for AddOrUpdateEndpointsAsync() method.
* Added validation of request to ensure that it does not contain two entries for the same endpoint.

karok2m committed:
Merge branch 'main' into feature/dmapi

karok2m committed:
Merge branch 'main' into feature/dmapi

karok2m committed:
Extended multi-endpoint E2E direct method tests. (#1549)

* Minor refactoring of direct method tests.
* Extended multi-endpoint test to check for all nodes.
* Changed cleanup to update published nodes files of both versions.
* Fixed expected node generation.
* Fixed multi-endpoint tests for legacy OPC Publisher.

karok2m committed:
Merge branch 'main' into feature/dmapi

karok2m committed:
Merge branch 'main' into feature/dmapi

monikavar committed:
Implement UnpublishAllNodes_V1 direct method (#1543)

* implement unpublishallnodes direct method

* adding unit tests and e2e tests

karok2m committed:
Merge branch 'main' into feature/dmapi

monikavar committed:
Add documentation for UnpublishAllNodes_V1 direct method. (#1553)

* add documentation for unpublishAllNodes direct method

* fix formatting of other direct methods

dacolgit committed:
Implemented GetDiagnosticInfo_V1 direct method. (#1544)

* getinfo start

* test

* test

* add GetDiagnosticInfoAsync dm

* remove test config files

* add unit test. fixes

* fix pr comments

* now GetDiagnosticInfo is called in DiagnosticsOutputTimer_Elapsed
fix pr comments

* remove input parameter from getDiagnosticInfo dm
fix pr comments

* fix conflict

* add  clear of _publisherDiagnosticInfo

* add GetDiagnosticInfo user doc
small fix

* fix

cristipogacean committed:
add support for incremental publisher reconfig (#1545)

* add support for incremental publisher reconfig
* cosmetic changes
* fix a a few lgtm reported issues
* fix issue with subscription name conflicts
* logs cosmetics on subscription services class
* review feedback
* bugfix

cristipogacean committed:
Publish/unpublish/addorpupdate/removeall nodes  empty response (#1554)

* the publish/ubpublish/addorpupdate/removeall nodes has empty response payload in case of success
* documentation update
* cosmetics

marcschier commented on components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs:
```diff
@@ -4,91 +4,88 @@
     /// Job orchestrator the represents the legacy publishednodes.json with legacy command line arguments as job.
     /// </summary>
-    public class LegacyJobOrchestrator : IJobOrchestrator {
+    public class LegacyJobOrchestrator : IJobOrchestrator, IPublisherConfigServices, IDisposable {
```
Eventually we should call this "StandaloneJobOrchestrator" or "LocalJobOrchestrator"

karok2m replied:
Renamed everything legacy to standalone.

marcschier commented on components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs:
```diff
@@ -141,173 +162,1014 @@ public class LegacyJobOrchestrator : IJobOrchestrator {
+            if (string.IsNullOrEmpty(content)) {
+                return null;
+            }
+            var sha = new SHA256Managed();
```
Check if this is disposable and use using

karok2m replied:
Fixed to use `using`.

karok2m committed:
Fixing LGTM warnings.

cristipogacean committed:
Renamed LegacyJobOrchestrator to StandaloneJobOrchestrator. (#1557)

* Remaned LegacyJobOrchestrator to StandaloneJobOrchestrator.
* Cosmetic changes.

Co-authored-by: Karapet Kostandyan <kakostan@microsoft.com>

cristipogacean commented on api/src/Microsoft.Azure.IIoT.OpcUa.Api.Publisher/src/Models/PublishedNodeApiModel.cs:
```diff
@@ -0,0 +1,98 @@
+        public string DisplayName { get; set; }
+
+        /// <summary>
+        /// Heartbeat interval as TimeSpan.
```
>Heartbeat interval as TimeSpan [](http://example.com/codeflow?start=12&length=30)

in seconds

cristipogacean replied:
fix in a next commit

cristipogacean commented on common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs:
```diff
@@ -28,6 +28,7 @@ public class MethodCallStatusException : MethodCallException {
         /// <param name="errorMessage"></param>
         public MethodCallStatusException(int result, string errorMessage = null) :
             this("{}", result, errorMessage) {
+            ResponsePayload = $"\"{errorMessage}\"";
```
>ResponsePayload [](http://example.com/codeflow?start=12&length=15)

change to json encodable object

cristipogacean replied:
fix in a next commit

cristipogacean commented on common/src/Microsoft.Azure.IIoT.Agent.Framework/src/Agent/Default/Worker.cs:
```diff
@@ -407,47 +459,49 @@ private class JobProcess : IDisposable {
+                    .ConfigureAwait(false);
 
                 // Process instructions
                 switch (result.HeartbeatInstruction) {
```
>switch [](http://example.com/codeflow?start=16&length=6)

add a default

cristipogacean replied:
fix in a next commit

cristipogacean reviewed (dismissed):
:shipit:

karok2m committed:
Fixing standalone E2E tests. (#1558)

* Changed mount directories for both versions of OPC Publisher to use subdirectories in /mount. This allows us to make sure that the directories exist before writing to published_nodes.json files.
* Making sure that we are in standalone mode before waiting for OPC Publisher module to be deployed.

cristipogacean reviewed (dismissed):
:shipit:

dacolgit committed:
fix removing of diagnosticInfo after an unpublish diect method call (#1559)

karok2m committed:
Using SSH client to create mount directories. (#1560)

Using SSH client instead of SFTP client to create mount directories for standalone E2E tests.

karok2m committed:
Fixing removal of diagnostics info in AddOrUpdateEndpointsAsync(). (#1561)

Aligning diagnostics info removal logic in AddOrUpdateEndpointsAsync() with that of UnpublishNodesAsync() and UnpublishAllNodesAsync().

karok2m committed:
Merge branch 'main' into feature/dmapi

cristipogacean committed:
Merge branch 'main' into feature/dmapi

karok2m committed:
Added E2E tests for AddOrUpdateEndpoints_V1 direct method. (#1563)

* Using AddOrUpdateEndpoints_V1 direct method in existing E2E tests.
* Cosmetic: UnPublishNodes -> UnpublishNodes
* Added AddOrUpdateEndpointsTest().
* Minor instrumentation in TestEventProcessor.
* Enabled full featured message for OPC Publisher 2.8.x tests.
* Added node unpublishing and validation.
* Fixing location of JobDiagnosticInfoModel so that it does not interfere with the main solution.
* Fixing OpcPublishingInterval assignment bug in ToApiModel() method.
* Fixing bug in test that used wrong response for diagnostic info.
* Using only date in names of deployments to make fewer of them in IoT Hub.
* Added incremental removal of nodes in AddOrUpdateEndpointsTest().
* ThresholdValue should not be 0.
* Waiting some time to allow test event processor to start.
